### ISSUE TITLE
Why use Pimple here?

### ISSUE THREAD

**Issue by GaryJones** [Rarst/wps#2]
What advantage does https://github.com/Rarst/wps/blob/master/php/class-plugin.php have by extending Pimple? Something inside the class? Something outside that helps other developers reference it (for unhooking etc.)?

**Comment by Giuseppe-Mazzapica** [comment]
# Short answer

Pimple is a DI container. Using it Rarst was able to:

- profit of the benefits of defining object as *services* : define once, obtain same instance everywhere
- profit of the benefit that Pimple factory callback brings: thanks to the fact they receive container instance as argument, it's easy build instance of objects, because through container is possible access to all other registered services
- manage dependency injection in a way that any consumer code can replace any of the objects used in the plugin with custom implementations
- benefit of the Pimple feaure that services are instantiated only if required: never used services are never instantiated

All these benefits come with no code, just estending Pimple.

# Long answer

Well, Pimple is a DI container. By extending it a class becomes able to store and provide services.

As example, inside Plugin class constructor you can see:

```php
$defaults['run'] = function ( $plugin ) {
   $run = new Run();
   $run->pushHandler( $plugin['handler.pretty'] );
   $run->pushHandler( $plugin['handler.json'] );
   return $run;
};
```

So, a **function** is stored in the container using the key `'run'` (Pimple implements [`ArrayAccess`](http://php.net/manual/en/class.arrayaccess.php)).

But when in the code there is ([here](https://github.com/Rarst/wps/blob/master/php/class-plugin.php#L114)):

```php
$run = $this['run'];
```
the variable `$run` is not a function, but it is an instance of `Run` class, i.e. it is the *result* of that function call.

So, you store functions, but you retrieve services (objects). This is the reason why those functions are named "factories".

I said factories return "services"  and not simply "objects" because you always retrieve same instance.

E.g.

```php
$run1 = $this['run'];
$run2 = $this['run'];
return $run1 === $run2; // true: same instance
```
Note that factories run only if services are ever required, otherwise they never run, improving performance.

The other great benefit in using Pimple is that when any factory function runs, it *magically* receives an instance of the container itself (in example above the `$plugin` variable). It means that inside any factory function you have access to all the other services, that allow you to easily put in place a dependency injection system.

For example, in the example above, the factory that creates the instance of `Run` contains:

```
$run->pushHandler( $plugin['handler.pretty'] );
```

`pushHandler` method receives `$plugin['handler.pretty']` that, [as you can see](https://github.com/Rarst/wps/blob/master/php/class-plugin.php#L59-L70), is an instance of another object defined in constructor.

By defining all the objects in constructor Rarst allows consumer code to easily overwrite any of the services, simply passing another factory callback to constructor with the same id, because default definitions [are merged with any definions passed to the constructor](https://github.com/Rarst/wps/blob/master/php/class-plugin.php#L88).

This is a pretty common practice, I've used it tons of times, e.g. in [my fork of this plugin](https://github.com/Giuseppe-Mazzapica/WPS/blob/master/src/Container.php), however, recently I realised that there are better ways to do same thing. If you have 5 spare minutes in [this blog post](http://gm.zoomlab.it/2014/structured-applications-with-pimple/) I explain which they are.

**Comment by GaryJones** [comment]
Thanks @Giuseppe-Mazzapica - I value your time and insight.

I (mostly) get DI containers, and I've been I've read up what I can about Pimple. I've used your Faber in a live project. I'd also read your blog post tonight, and am currently in the middle of creating my own plugin boilerplate using your notes there.

What I'd clearly missed in every way (or mentally skipped over due to unfamiliarity and forget to go back and check) was the setting up of Pimple with services in the constructor. You've highlighted that, and it's now very obvious.

Thanks again for your time - I really appreciate it, as I do want to really get to grips with this.

**Comment by Rarst** [comment]
TL;DR extending Pimple in a class allows it to act as container and implement set of default predefined services.

For the record Pimple got services relatively recently, so my projects mostly predate that. I had to come up with my own convention for handler registrations in Fragment Cache at the time.

**Comment by Giuseppe-Mazzapica** [comment]
Yep. Pimple got service provider on version 3 (24 Jul 2014). And I remember that in older version docs, suggested way for custom Containers was extending Pimple.

The code in the older docs was pretty the same I used in [Brain](https://github.com/Giuseppe-Mazzapica/Brain/blob/master/Container.php#L12). Apart from `brain_modules`: that was my way to do service providers...
